stm32-modules/heater-shaker/firmware/heater_task/heater_hardware.c

```diff
@@ -285,39 +288,47 @@ void heater_hardware_power_disable(heater_hardware* hardware) {
     }
 }
 
-bool heater_hardware_power_set(heater_hardware* hardware, uint16_t setting) {
+HEATPAD_CIRCUIT_ERROR heater_hardware_power_set(heater_hardware* hardware, uint16_t setting) {
     hw_internal* internal = (hw_internal*)hardware->hardware_internal;
     if (!internal) {
         init_error();
     }
     if (heatpad_in_error_state()) {
-        return false;
+        if (internal->heatpad_cs_status == ERROR_SHORT_CIRCUIT) {
+            return SHORT;
+        } else if (internal->heatpad_cs_status == ERROR_OPEN_CIRCUIT) {
+            return OPEN;
+        } else if (internal->heatpad_cs_status == ERROR_OVERCURRENT) {
+            return OVERCURRENT;
+        }
     }
     if (internal->heatpad_cs_status == IDLE) {
         internal->heatpad_cs_status = RUNNING;
     }
-    internal->pwm_pulse_duration = setting;
-    internal->pwm_config.Pulse = setting;
-    if (!internal->heater_started) {
-        HAL_TIM_PWM_Stop_IT(&internal->pad_tim, HEATER_PAD_ENABLE_TIM_CHANNEL);
-        HAL_TIM_PWM_ConfigChannel(
-            &internal->pad_tim, &internal->pwm_config, HEATER_PAD_ENABLE_TIM_CHANNEL);
-        HAL_TIM_PWM_Start_IT(&internal->pad_tim, HEATER_PAD_ENABLE_TIM_CHANNEL);
-        internal->pwm_config.Pulse = HEATER_PAD_SHORT_CHECK_PULSE;
-        HAL_TIM_PWM_Stop_IT(&internal->pad_tim, HEATER_PAD_SHORT_CHECK_TIM_CHANNEL);
-        HAL_TIM_PWM_ConfigChannel(
-            &internal->pad_tim, &internal->pwm_config, HEATER_PAD_SHORT_CHECK_TIM_CHANNEL);
-        HAL_TIM_PWM_Start_IT(&internal->pad_tim, HEATER_PAD_SHORT_CHECK_TIM_CHANNEL);
-        internal->pwm_config.Pulse = HEATER_PAD_OPEN_CHECK_PULSE;
-        HAL_TIM_PWM_Stop_IT(&internal->pad_tim, HEATER_PAD_OPEN_CHECK_TIM_CHANNEL);
-        HAL_TIM_PWM_ConfigChannel(
-            &internal->pad_tim, &internal->pwm_config, HEATER_PAD_OPEN_CHECK_TIM_CHANNEL);
-        HAL_TIM_PWM_Start_IT(&internal->pad_tim, HEATER_PAD_OPEN_CHECK_TIM_CHANNEL);
-        internal->heater_started = true;
-    } else {
-        HEATER_PAD_LL_SETCOMPARE(internal->pad_tim.Instance, setting);
+    if (!internal->update_lock) {
```

Comment:
My understanding is that, if the update lock is enabled, this function will silently return without updating the power. If we need to lock out PWM updates for period of time, I think it makes sense to actually cache the setting and then update the timer compare value once the lock is cleared.

Reply:
<img width="676" alt="ramp with cached_pulse_setting" src="https://user-images.githubusercontent.com/25947011/180277557-e8ca0dd1-1985-42a8-bfc8-b385b5cc85bc.png">

Reply:
<img width="661" alt="ramp without cached_pulse_setting" src="https://user-images.githubusercontent.com/25947011/180277718-0490aa04-f913-408f-b47b-f1e02a41fc84.png">
ramp without cached_pulse_setting

Reply:
Not using the cached_pulse_setting will prevent pwm pulse from updating for one cycle. The following cycle, the PID will calculate power update with the new, increased error. I think this approach seems correct, and the ramp profile seems to agree